packages/python/accern_data/accern_data.py

```diff
@@ -574,74 +627,119 @@ def download_range(
-        else:
-            valid_filters = self.parse_filters(
-                {**self.get_filters(), **self.validate_filters(filters)})
-        if end_date is None:
-            expected_records.append(
-                self._read_total(start_date, valid_filters))
-            print_fn(f"single day {start_date}")
-            print_fn(f"expected {expected_records[0]}")
-            progress_bar = ProgressBar(
-                    total=expected_records[0],
-                    desc="Downloading signals",
-                    verbose=verbose)
-            self._process_date(
-                start_date,
-                output_path,
-                output_pattern,
-                is_first_time=True,
-                mode=valid_mode,
-                filters=valid_filters,
-                progress_bar=progress_bar)
-        else:
-            is_first_time = True
-            total = 0
-            progress_bar = ProgressBar(
-                total=len(pd.date_range(start_date, end_date)),
-                desc="Fetching info",
-                verbose=verbose)
-
-            for cur_date in pd.date_range(start_date, end_date):
-                expected_records.append(
-                    self._read_total(cur_date, valid_filters))
-                progress_bar.update(1)
-
-            total = sum(expected_records)
-            progress_bar.set_total(total=total)
-            progress_bar.set_description(desc="Downloading signals")
-
-            for ix, cur_date in enumerate(pd.date_range(start_date, end_date)):
-                print_fn(f"now processing {cur_date}")
-                print_fn(f"expected {expected_records[ix]}")
-                progress_bar.set_description(
-                    f"Downloading signals for {cur_date.strftime('%Y-%m-%d')}")
-                is_first_time = self._process_date(
-                    cur_date.strftime("%Y-%m-%d"),
+
+        valid_mode: Optional[Mode] = None
+        cur_date: Optional[pd.Timestamp] = None
+        prev_date: Optional[pd.Timestamp] = None
+        indicator_obj: Optional[ProgressIndicator] = None
+
+        def set_active_mode(
+                mode: Mode,
+                date: pd.Timestamp,
+                indicator: ProgressIndicator) -> None:
+            nonlocal valid_mode, cur_date, indicator_obj
+            valid_mode = mode
+            cur_date = date
+            indicator_obj = indicator
+
+        is_first_time = True
+        for res in self.iterate_range(
+                start_date=start_date,
+                end_date=end_date,
+                mode=mode,
+                filters=filters,
+                chunk_size=None,
+                indicator=indicator,
+                set_active_mode=set_active_mode):
+            assert valid_mode is not None
+            assert cur_date is not None
+            assert indicator_obj is not None
+
+            if cur_date != prev_date:
+                if prev_date is not None:
```

Comment:
hmm, I'm thinking -- you could do the finish_day and init_day calls in the set_active_mode function directly then the only thing you need to do in the actual loop is add_result and prev_date

Reply:
no, `set_active_mode` gets called even before we have the data. so there is no point calling `finish_day` or `init_day`

Reply:
I'm talking about calling those functions inside the set_active_mode function. this function should get called for every new day so that is the best place to call init_day and finish_day